## KPMG Virtual Analytics Internship Program
# Starting of Data Analysis Proccess (EDA - Exploratory Data Analysis)

In [1]:
import pandas as pd
sheetLists = [1,3,4]
dfList = []
for number in sheetLists:
    dfList.append(pd.read_excel('KPMG_VI_New_raw_data_update_final.xlsx', sheet_name=number,header=1))
    
# transactions_df = pd.read_excel('KPMG_VI_New_raw_data_update_final.xlsx', sheet_name=1,header=1)
# cust_demographics_df = pd.read_excel('KPMG_VI_New_raw_data_update_final.xlsx', sheet_name=3, header=1)
# cust_address = pd.read_excel('KPMG_VI_New_raw_data_update_final.xlsx', sheet_name=4, header=1)

#First entry will be trasnactions in the array, second customer demographis and third customer address. 

C:\Users\WBurc\AppData\Local\Temp\ipykernel_25852\2913502600.py:5: FutureWarning: Inferring datetime64[ns] from data containing strings is deprecated and will be removed in a future version. To retain the old behavior explicitly pass Series(data, dtype=datetime64[ns])
  dfList.append(pd.read_excel('KPMG_VI_New_raw_data_update_final.xlsx', sheet_name=number,header=1))


After loading the data, we will quickly confirm the headers are in proper and data appears roughly as expected.

In [2]:
#cycling through the three dataframes to insure header/data output appears correctly.
for x in dfList:
    print(x.head(2))

   transaction_id  product_id  customer_id transaction_date  online_order  \
0           16701           0         5034       2017-01-27           0.0   
1            8708           0         5034       2017-10-07           0.0   

  order_status           brand product_line product_class product_size  \
0     Approved  Norco Bicycles     Standard        medium       medium   
1     Approved           Solex         Road        medium       medium   

   list_price  standard_cost  product_first_sold_date  
0      360.40     270.299988                  35667.0  
1      416.98     312.735016                  42218.0  
   customer_id first_name last_name  gender  \
0         1092     Katlin   Creddon  Female   
1         3410    Merrili   Brittin  Female   

   past_3_years_bike_related_purchases        DOB           job_title  \
0                                   56 1935-08-22  VP Quality Control   
1                                   93 1940-09-22                 NaN   

  job_industry_

Next, we can quickly examine our tables with the .info() command which will tell us the data type of each column and the count of non-null values. We can also see the number of missing entries summed by column with a single command using Jypter Notebook/Pandas.

In [3]:
#cycling through the three dataframes to check for data inconsistency 
#Beginning with df(0) or transactions
dfList[0].info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 20000 entries, 0 to 19999
Data columns (total 13 columns):
 #   Column                   Non-Null Count  Dtype         
---  ------                   --------------  -----         
 0   transaction_id           20000 non-null  int64         
 1   product_id               20000 non-null  int64         
 2   customer_id              20000 non-null  int64         
 3   transaction_date         20000 non-null  datetime64[ns]
 4   online_order             19640 non-null  float64       
 5   order_status             20000 non-null  object        
 6   brand                    19803 non-null  object        
 7   product_line             19803 non-null  object        
 8   product_class            19803 non-null  object        
 9   product_size             19803 non-null  object        
 10  list_price               20000 non-null  float64       
 11  standard_cost            19803 non-null  float64       
 12  product_first_sold_date  19803 n

Most columns has 20,000 values in transcations except "order online" that is missing 360 values, and the columns "brand, product_line, product_class, product_size, standard_cost and product_first_sold_date" which is missing 197 entries.

In [4]:
dfList[0].isnull().sum()

transaction_id               0
product_id                   0
customer_id                  0
transaction_date             0
online_order               360
order_status                 0
brand                      197
product_line               197
product_class              197
product_size               197
list_price                   0
standard_cost              197
product_first_sold_date    197
dtype: int64

Moving onto the customer demographics table

In [5]:
dfList[1].info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 4000 entries, 0 to 3999
Data columns (total 13 columns):
 #   Column                               Non-Null Count  Dtype         
---  ------                               --------------  -----         
 0   customer_id                          4000 non-null   int64         
 1   first_name                           4000 non-null   object        
 2   last_name                            3875 non-null   object        
 3   gender                               4000 non-null   object        
 4   past_3_years_bike_related_purchases  4000 non-null   int64         
 5   DOB                                  3913 non-null   datetime64[ns]
 6   job_title                            3494 non-null   object        
 7   job_industry_category                3344 non-null   object        
 8   wealth_segment                       4000 non-null   object        
 9   deceased_indicator                   4000 non-null   object        
 10  default     

We can see job_industry_category has the most missing entries, followed by job_title and default. Overall, there are a total of 400 records. The next command sum the number of null (empty) values in our data. 125 customers are missing last names. 87 customers have no date of birth. Job titles and categories are missing approximately 12.5% and 16.5% of the time respectively. Default appears to be missing 302 times and when examining this column, it largely seems to contain junk-data and may be corrupted data. Sample entries include 'Åâ´â°ËèËÃâââ' and 'img src=x onerror=alert('hi') />'

In [6]:
dfList[1].isnull().sum()

customer_id                              0
first_name                               0
last_name                              125
gender                                   0
past_3_years_bike_related_purchases      0
DOB                                     87
job_title                              506
job_industry_category                  656
wealth_segment                           0
deceased_indicator                       0
default                                302
owns_car                                 0
tenure                                  87
dtype: int64

Looking at our third table, we can see no missing entries.

In [7]:
dfList[2].info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 3999 entries, 0 to 3998
Data columns (total 6 columns):
 #   Column              Non-Null Count  Dtype 
---  ------              --------------  ----- 
 0   customer_id         3999 non-null   int64 
 1   address             3999 non-null   object
 2   postcode            3999 non-null   int64 
 3   state               3999 non-null   object
 4   country             3999 non-null   object
 5   property_valuation  3999 non-null   int64 
dtypes: int64(3), object(3)
memory usage: 187.6+ KB


In [14]:
dfList[2].isnull().sum()

customer_id           0
address               0
postcode              0
state                 0
country               0
property_valuation    0
dtype: int64

While there 4000 customer demograpic entries, there are only 3999 customer addresses listed. There must be at least 1 duplicate demographic entries or 1 missing customer address.

# Summary of Data Contained In The Transaction Data Sheet

In [15]:
dfList[0].describe()

transaction_id   product_id   customer_id  online_order    list_price  \
count    20000.000000  20000.00000  20000.000000  19640.000000  20000.000000   
mean     10000.500000     45.36465   1738.246050      0.500458   1107.829449   
std       5773.647028     30.75359   1011.951046      0.500013    582.825242   
min          1.000000      0.00000      1.000000      0.000000     12.010000   
25%       5000.750000     18.00000    857.750000      0.000000    575.270000   
50%      10000.500000     44.00000   1736.000000      1.000000   1163.890000   
75%      15000.250000     72.00000   2613.000000      1.000000   1635.300000   
max      20000.000000    100.00000   5034.000000      1.000000   2091.470000   

       standard_cost  product_first_sold_date  
count   19803.000000             19803.000000  
mean      556.046951             38199.776549  
std       405.955660              2875.201110  
min         7.210000             33259.000000  
25%       215.140000             35667.000000  
50%       507.580000             38216.000000  
75%       795.100000             40672.000000  
max      1759.850000             42710.000000

Looking at the output of the describe feature for transaction dataframe, we can see the average list price of items sold was $582.82 and 50.05% of sales were online or nearly half. The least expensive item sold was for $12.01. The most expensive item sold was for $2091.47. The median price, where exactly half the items sold for less and half the items sold for more was $1163.89. While one quarter was under $575.27 and one quarter was above $1163.89.

Another observation is list price is not listed as a currency in the excel file while standard cost is listed as a currency. Reformatting the row might be better for your data clarity.

In [16]:
print("Customer_ID statistics of Transaction dataframe: \n")
print("Count of Customer_Ids: ", dfList[0]['customer_id'].count(),"\nUnique customer IDs: ", len(dfList[0]['customer_id'].unique()), "\nLargest Customer ID: ", dfList[0]['customer_id'].max())

Customer_ID statistics of Transaction dataframe: 

Count of Customer_Ids:  20000 
Unique customer IDs:  3494 
Largest Customer ID:  5034


From this data, we can see there are 20,000 UserIDs listed, 3494 unique IDs and 5034 is the maximum ID. It is likely 5034 being so far above the other largest customer IDs is a data entry error. Ex. Perhaps 0534 or 3405 was intended. The maximum userID jumps from 3500 to 5034 and seems to abruptly change. This customer's sale data also contains standard prices with an unusual quality of decimal places not standard for a currency figure.

In [17]:
dfList[0][dfList[0].standard_cost > dfList[0].list_price]

Empty DataFrame
Columns: [transaction_id, product_id, customer_id, transaction_date, online_order, order_status, brand, product_line, product_class, product_size, list_price, standard_cost, product_first_sold_date]
Index: []

We can also confirm there are no instances were an item was sold at a loss. Next, we'll check for null values and determine the percentage of entries with null values. We can see there are 360 values where it is unclear of the order was online or in person and 197 orders for which the brand, product_line, product_class, product_size, standard_cost and product_first_sold_date is absent. This represents approximate 0.985% to 1.8% of the data. If we were to remove this data from our analysis, we could perform analytics based upon over 98% of the customer data.

In [25]:
(dfList[0].isnull().sum() / 20000) * 100

transaction_id             0.000
product_id                 0.000
customer_id                0.000
transaction_date           0.000
online_order               1.800
order_status               0.000
brand                      0.985
product_line               0.985
product_class              0.985
product_size               0.985
list_price                 0.000
standard_cost              0.985
product_first_sold_date    0.985
dtype: float64

In [26]:
dfList[0].isnull().sum()

transaction_id               0
product_id                   0
customer_id                  0
transaction_date             0
online_order               360
order_status                 0
brand                      197
product_line               197
product_class              197
product_size               197
list_price                   0
standard_cost              197
product_first_sold_date    197
dtype: int64

# Summary of Data Contained In The Customer Demographic Data Sheet

Examining the statistical data for this table, we can see over the last three years, the average customer purchased about 49 bike repeated purchases. The maximum purchases any given customer made was 99 and the minimum was 0. A few columns could be better described. The column Tensure is slightly unclear. Was this meant the duration a person has been a customer of our client or the time the customer has worked at their current employer? Wealth Segment is also a categorial variable that might be more useful to have replaced with a number-range for calculations. There are currently three-different categories. "Mass Customer", "Affluent Customer" and "High Networth Customer". What is the difference in wealth segement between having a high networth and being affluent? The category leaves this somewhat unclear.
341770 341770 341770

In [27]:
dfList[1].describe()

customer_id  past_3_years_bike_related_purchases       tenure
count  4000.000000                          4000.000000  3913.000000
mean   2000.500000                            48.890000    10.657041
std    1154.844867                            28.715005     5.660146
min       1.000000                             0.000000     1.000000
25%    1000.750000                            24.000000     6.000000
50%    2000.500000                            48.000000    11.000000
75%    3000.250000                            73.000000    15.000000
max    4000.000000                            99.000000    22.000000

Checking for null (missing) values and calculating the percentages, we can see significantly more data is missing from our demographics table by percentage. We can see there are 360 values where it is unclear of the order was online. This figure matches the quantity of the missing data in the transaction table. 197 orders for which the brand, product_line, product_class, product_size, standard_cost and product_first_sold_date is absent. 16.4% of job industry categories and 12.65% of job titles are missing from our dataset. The category "default" containing junk-data is the third most common missing category. The other columns have above 96% completion. If we were to drop these customers, we could perform analytics on 96% of the dataset. If customer industry an important criteria for your business, you wish for analytics to be calculated on, 83.6% is the maximum customer base percentage we could use.

In [31]:
(dfList[1].isnull().sum() / 4000) * 100

customer_id                             0.000
first_name                              0.000
last_name                               3.125
gender                                  0.000
past_3_years_bike_related_purchases     0.000
DOB                                     2.175
job_title                              12.650
job_industry_category                  16.400
wealth_segment                          0.000
deceased_indicator                      0.000
default                                 7.550
owns_car                                0.000
tenure                                  2.175
dtype: float64

In [32]:
dfList[1].isnull().sum()

customer_id                              0
first_name                               0
last_name                              125
gender                                   0
past_3_years_bike_related_purchases      0
DOB                                     87
job_title                              506
job_industry_category                  656
wealth_segment                           0
deceased_indicator                       0
default                                302
owns_car                                 0
tenure                                  87
dtype: int64